core/runtime.rs

```diff
@@ -1111,89 +1115,81 @@ impl JsRuntime {
   fn evaluate_pending_module(&mut self) {
     let state_rc = Self::state(self.v8_isolate());
 
-    let context = self.global_context();
-    {
-      let scope =
-        &mut v8::HandleScope::with_context(self.v8_isolate(), context);
+    let maybe_module_evaluation =
+      state_rc.borrow_mut().pending_mod_evaluate.take();
 
-      let mut state = state_rc.borrow_mut();
+    if maybe_module_evaluation.is_none() {
+      return;
+    }
 
-      if let Some(module_evaluation) = state.pending_mod_evaluate.as_ref() {
-        let promise = module_evaluation.promise.get(scope);
-        let mut sender = module_evaluation.sender.clone();
+    let module_evaluation = maybe_module_evaluation.unwrap();
+    let scope = &mut self.handle_scope();
+
+    let promise = module_evaluation.promise.get(scope);
+    let mut sender = module_evaluation.sender.clone();
+    let promise_state = promise.state();
+
+    match promise_state {
+      v8::PromiseState::Pending => {
+        // NOTE: `poll_event_loop` will decide if
+        // runtime would be woken soon
+        state_rc.borrow_mut().pending_mod_evaluate = Some(module_evaluation);
```

Comment:
It seems odd to me that you call `pending_mod_evaluate.take()` only to return the value here. Is there something wrong with the previous `if let Some(...` construction?

Reply:
There were previously two calls to `pending_mod_evaluate.take()`, I believe this version is simpler